### PR TITLE
Add EXPLAIN (plan).

### LINKED ISSUE #2172: sql: add support for EXPLAIN (non-debug)
Now that we have a `sortNode` and a `scanNode`, an `EXPLAIN` statement that shows the query plan is interesting and would allow further verification in logic tests that we're doing what we expect (i.e. not sorting when sorting is not required).

### PR DESCRIPTION
**Pull request by petermattis**
The output is pretty redumentary, containing Level, Type and Description
field. The Level field indicates the level in the query plan tree (with
0 indicating the root). The Type indicates the type of the plan
node (currently scan, revscan, sort, nosort, or values). The Description
is some node specificy information, such as the table/index being
scanned or the keys being sorted on.

Fixed a buglet in valuesNode: the values lists must all be the same
length.

Fixes #2172.

### CODE REVIEW

File: sql/explain.go [hunk 1]
```diff
@@ -48,9 +51,20 @@ func (p *planner) Explain(n *parser.Explain) (planNode, error) {
...
+			return nil, fmt.Errorf("%v: %s", err, n)
+		}
+		return plan, nil
+	case explainPlan:
+		v := &valuesNode{}
+		v.columns = []string{"Level", "Type", "Description"}
+		explainNode(v, plan, 0)
+		plan = v
+	default:
+		return nil, fmt.Errorf("unsupported EXPLAIN mode: %d", mode)
```
Comment (tamird) [inline sql/explain.go]: unreachable?

File: sql/explain.go [hunk 2]
```diff
@@ -48,9 +51,20 @@ func (p *planner) Explain(n *parser.Explain) (planNode, error) {
 	if err != nil {
 		return nil, err
 	}
-	plan, err = markExplain(plan, mode)
-	if err != nil {
-		return nil, fmt.Errorf("%v: %s", err, n)
+	switch mode {
+	case explainDebug:
+		plan, err = markExplain(plan, mode)
```
Comment (tamird) [inline sql/explain.go]: the naming is getting a bit fuzzy with `explainNode`. perhaps a good time to rename

File: sql/explain.go (hunk 1 above)
Comment (petermattis) [inline sql/explain.go]: Yes, but only because we exit earlier if `mode == explainNone`. Seems safer to leave this in in case a new mode gets added without enhancing this switch.

File: sql/explain.go (hunk 2 above)
Comment (petermattis) [inline sql/explain.go]: Sure. Do you have a better name?

Comment (tamird) [inline sql/explain.go]: I'd name this one `markDebug` since it really behaves more like a normal select with the debug option, and I'd change `explainNode` -> `populateExplain` and possibly make it a method on `valuesNode`

Comment (petermattis) [inline sql/explain.go]: `markExplainDebug`?

Comment (petermattis) [inline sql/explain.go]: Ok, I did the `markDebug` and `populateExplain` renames, but I resisted making `populateExplain` a method on `valuesNode`. Explaining a plan really seems unrelated to the rest of the `valuesNode` functionality.

File: sql/plan.go [hunk 3]
```diff
@@ -136,6 +136,9 @@ type planNode interface {
 	Next() bool
 	// Err returns the error, if any, encountered during iteration.
 	Err() error
+	// ExplainPlan fills in name and description for the node and returns a list
+	// of child nodes.
+	ExplainPlan(name, description *string) []planNode
```
Comment (tamird) [inline sql/plan.go]: is this really better than returning the strings?

File: sql/explain.go [hunk 4]
```diff
@@ -64,9 +78,24 @@ func markExplain(plan planNode, mode explainMode) (planNode, error) {
...
 	default:
 		return nil, fmt.Errorf("TODO(pmattis): unimplemented %T", plan)
 	}
 }
+
+func populateExplain(v *valuesNode, plan planNode, level int) {
+	var name, description string
+	children := plan.ExplainPlan(&name, &description)
+
+	row := make(parser.DTuple, len(v.columns))
```
Comment (tamird) [inline sql/explain.go]: nit: this can just be a `row` literal?

File: sql/values.go [hunk 5]
```diff
@@ -37,8 +39,18 @@ func (p *planner) Values(n parser.Values) (planNode, error) {
 		if !ok {
 			return nil, fmt.Errorf("expected a tuple, but found %T", data)
 		}
+		if len(v.rows) == 0 {
+			nCols = len(vals)
+		} else if nCols != len(vals) {
+			return nil, fmt.Errorf("VALUES lists must all be the same length")
```
Comment (tamird) [inline sql/values.go]: this can be instantiated just once with `errors.New`

File: sql/plan.go (hunk 3 above)
Comment (petermattis) [inline sql/plan.go]: I'm not sure. I actually started off returning strings and switched to this. If we add more detail to `EXPLAIN` we might want to actually return an `explainPlanInfo` struct. Do you care strongly about this now?

File: sql/explain.go (hunk 4 above)
Comment (petermattis) [inline sql/explain.go]: `parser.DTuple` has the type `[]parser.Datum`. I'm allocating a slice to a particular length, how can that be a literal?

File: sql/values.go [hunk 6]
```diff
@@ -37,8 +39,18 @@ func (p *planner) Values(n parser.Values) (planNode, error) {
...
+		}
 		v.rows = append(v.rows, vals)
 	}
+
+	v.columns = make([]string, nCols)
+	for i := 0; i < nCols; i++ {
+		v.columns[i] = fmt.Sprintf("column%d", i)
```
Comment (tamird) [inline sql/values.go]: i can't match this up to anything in the tests; what does it do?

File: sql/plan.go (hunk 3 above)
Comment (tamird) [inline sql/plan.go]: I feel like out-parameters is a pretty exotic choice in Go, so I'd rather avoid it. It doesn't buy clarity since the caller still has to assign names to positional arguments, so I'd rather we used multiple return here.

File: sql/values.go (hunk 5 above)
Comment (petermattis) [inline sql/values.go]: Switched to `errors.New` but I'm hesitant to place it in a global var. Doing so seems like premature optimization for a message that is used once. IMO, it is better to keep the error message next to the code where the error is generated.

File: sql/explain.go (hunk 4 above)
Comment (tamird) [inline sql/explain.go]:
```go
row := parser.DTuple{
  parser.DInt(level),
  parser.DString(name),
  parser.DString(description),
}
```

File: sql/values.go (hunk 6 above)
Comment (petermattis) [inline sql/values.go]: It gives a `VALUES` statement output column names. I think I accidentally deleted the test case for it. Without column names we pass invalid data back to the client (length of the column names != the number of columns per row).

File: sql/values.go (hunk 5 above)
Comment (tamird) [inline sql/values.go]: this error also needs a test case, yeah?

Comment (petermattis) [inline sql/values.go]: Good point. Added.

File: sql/explain.go (hunk 4 above)
Comment (petermattis) [inline sql/explain.go]: Oh, good point. Duh.

File: sql/testdata/values [hunk 7]
```diff
@@ -0,0 +1,10 @@
+query ITT colnames
+VALUES (1, 2, 3), (4, 5, 6)
+----
+column1 column2 column3
```
Comment (tamird) [inline sql/testdata/values]: :+1:

**Comment by tamird** [comment]
LGTM modulo out parameters

File: sql/plan.go (hunk 3 above)
Comment (petermattis) [inline sql/plan.go]: Switched to return values. Doesn't really affect the code much at all, but I suppose it is more idiomatic.
